ansible_ai_connect/ai/api/views.py

```diff
@@ -134,11 +128,100 @@
 }
 
 
-class Completions(APIView):
+class OurAPIView(APIView):
+    exception = None
+
+    def initial(self, request, *args, **kwargs):
+        super().initial(request, *args, **kwargs)
+        self.request = request
+        request_serializer = self.serializer_class(data=request.data, context={"request": request})
+        request_serializer.is_valid(raise_exception=True)
+        self.validated_data = request_serializer.validated_data
+        if self.schema1_event_class:
+            self.schema1_event = self.schema1_event_class.init(request.user, self.validated_data)
+
+    def _get_model_name(self, org_id: str) -> str:
+        try:
+            model_mesh_client = apps.get_app_config("ai").model_mesh_client
+            model_name = model_mesh_client.get_model_id(
+                org_id, self.validated_data.get("model", "")
+            )
+            return model_name
+        except (WcaNoDefaultModelId, WcaModelIdNotFound, WcaSecretManagerError):
+            return ""
+
+    def handle_exception(self, exc):
+        self.exception = exc
+
+        # Mapping between the internal exceptions and the API exceptions (with a message and a code)
+        mapping = [
+            (WcaInvalidModelId, WcaInvalidModelIdException),
+            (WcaBadRequest, WcaBadRequestException),
+            (WcaKeyNotFound, WcaKeyNotFoundException),
+            (WcaModelIdNotFound, WcaModelIdNotFoundException),
+            (WcaNoDefaultModelId, WcaNoDefaultModelIdException),
+            (WcaEmptyResponse, WcaEmptyResponseException),
+            (WcaCloudflareRejection, WcaCloudflareRejectionException),
+            (WcaUserTrialExpired, WcaUserTrialExpiredException),
+        ]
+
+        for original_class, new_class in mapping:
+            if isinstance(exc, original_class):
+                exc = new_class(cause=exc)
+                break
+        logger.error(f"{type(exc)}: {traceback.format_exception(exc)}")
+        response = super().handle_exception(exc)
+        return response
+
+    def get_ids(self):
+        allowed = ["explanationId", "generationId", "suggestionId"]
+        # Return the ids we want to include in the exception messages
+        ret = {}
+        for k, v in self.validated_data.items():
+            if k in allowed and v:
+                ret[k] = v
+            elif isinstance(v, dict):
+                for subk, subv in v.items():
+                    if subk in allowed and subv:
+                        ret[subk] = subv
+        return ret
+
+    def dispatch(self, request, *args, **kwargs):
+        start_time = time.time()
+        self.exception = False
+        self.schema1_event = None
+        response = super().dispatch(request, *args, **kwargs)
+
+        if self.schema1_event:
```

Comment:
If the request fails before `initial` is called, this won't be set. One example where this occurs is if authentication fails.

Reply:
Yes, this is an assumed decision. Those authentication errors should be handled differently IMO.